More colour, better error/warning messages

anjohnson opened the pull request:
Various changes to iocsh.cpp and others to add and improve the use of color.

- The IOC Shell output when running a startup script now displays comment lines in Cyan,
and Bold for command lines.
- The `softIoc -v` output also uses the same color scheme for the commands it shows.
- The default IOC Shell prompt is now green, but the choice of color can be edited in the
`configure/CONFIG_SITE_ENV` file – the default `ENV_PARAM` values can use the `ANSI_*()`
macros to select the text color.
- More error messages printed by the shell appear in Red, or use the red `ERROR` prefix.
- The word "Illegal" has been replaced with "Invalid" in several iocsh error messages.
- The iocsh `var` command now shows the data type of the registered variable.

![image](https://github.com/user-attachments/assets/d6a0b3a2-aa33-4ddd-8e1c-67a1794b4d75)

anjohnson committed:
Add more colour to iocsh output messages

Includes some unification of error text, narrower wrap margin.
'var' output now lists the variable types.

anjohnson committed:
Make default iocsh prompt green

The bldEnvData.pl script now syntax-checks the envParam
default values, allowing ANSI color macros to be used
in the IOCSH_PS1 parameter.

anjohnson committed:
Improve iocsh tests, update extected output

anjohnson committed:
Update & color error/warning messages from db/dbd parser

anjohnson committed:
Add color to the softIoc verbose output

anjohnson committed:
Use "Invalid" instead of "Illegal" in error messages

anjohnson committed:
asSet* commands don't need DBD loaded

mdavidsaver commented on modules/libcom/src/iocsh/iocsh.cpp:
```diff
@@ -927,9 +930,9 @@ static void helpCallFunc(const iocshArgBuf *args)
-                "Type 'help <command>' to see the arguments of <command>.  eg. 'help db*'\n");
+        fprintf(epicsGetStdout(), "\n"
+            "Type 'help <command>' to see the arguments of <command>, "
+            "eg. 'help db*'\n");
```
A minor point, but isn't it preferable for readability for a string constant with two newlines to split across 3 source lines?

anjohnson replied:
I was trying to use an 80 column text-wrap width, but given the `ANSI_UNDERLINE()` string below it that's probably too pedantic here. I will fix.

anjohnson replied:
The original wording doesn't mention glob patterns at all, and my fix for that grew to a second line anyway:
```patch
         fprintf(epicsGetStdout(), "\n"
-            "Type 'help <command>' to see the arguments of <command>, "
-            "eg. 'help db*'\n");
+            "Type 'help <glob>' for information about commands matching\n"
+            "the name or pattern <glob>, e.g. 'help db*'\n");
```

New commits coming soon.

mdavidsaver reviewed (approved):
I like it.

simon-ess commented:
Is it possible to configure this at all? It seems that at the moment the only configurable portion is the `PS1`; however we are stuck with the default coloration/highlighting for all other output.

(But I agree with @mdavidsaver  - I think that this is a nice addition)

mdavidsaver commented:
Codacy flags a "Variable 'line' was passed to a function after being freed.", which imo. is an obvious false positive.

anjohnson commented:
@mdavidsaver Thanks for confirming my thoughts about Codacy's report.

anjohnson commented:
I wondered if configurability would come up. This may show one reason it's desirable, although it doesn't seem quite as bad as I thought it might be:

![image](https://github.com/user-attachments/assets/a795b737-2c71-4b9c-8cdc-41780aa6bdf8)

Generalization of the `verbose_out()` function I added to `softIoc/softMain.cpp` might be a way to do that, it takes an enum value named after the type of text to be printed and looks up the ANSI color code for that in a table. Doing that across the code-base might be nice, but there are rather a lot of different kinds of ways that we provide and print messages, and I could see it being complicated to handle them all.

The `ANSI_ESC_` strings are mostly expanded out at compile-time, so they only add a few characters to the output and no additional levels of indirection/function calls.

mdavidsaver commented:
> I wondered if configurability would come up.

imo. schema control is a solid nice to have going forward, provided it does not add too much complexity.  I certainly do not see it as a blocker.

I see a reasonable contrast with both light on dark as well as dark on light color schemes as sufficient for this PR.  (which I do in fact see)

For #113 I did spend some time on contrast when selecting the colors.  Although I must admit that `ANSI_CYAN` is the hardest to read on a light background.  This is why I have PVXS use the darker `ANSI_BLUE` for info level log messages.

anjohnson committed:
Improve 'help' self-description message

anjohnson committed:
Add Release note (using new-notes scheme from PR #628)

anjohnson commented:
This image shows the use of `ANSI_BLUE` instead of `ANSI_CYAN` for comments:

<img width="764" alt="image" src="https://github.com/user-attachments/assets/a18a2ebf-f5b3-42cc-ab40-0160cd25ad14" />

I changed the code to use that instead, and added a release note. I think this PR is now ready (although Codacy still disagrees).

anjohnson commented:
Core Group: Configurability is desirable, but not essential to begin with (later PR). Use logical names for the things we configure, maybe re-use GNU's `LS_COLORS` and `GREP_COLORS` scheme. Consider removing the change to allow `ANSI_*` in `CONFIG_SITE_ENV`.

mdavidsaver commented:
A couple of thoughts...

wrt. scripts trying to parse stdout/stderr.  I think mainly tester scripts looking for the prompt string.  Here anything we do to change the default prompt will break a tester which assumes the current default.  I guess in this case our statement is that tester scripts should set `$IOCSH_PS1`?

I also find myself wondering if the `funopen()` / `fopencookie()` calls (cf. https://github.com/epics-base/epics-base/pull/248) might be useful as vectors for conditionally striping out escape sequences from stdout/err ?

anjohnson commented:
Rebased and made some final code adjustments.

In light of @mdavidsaver's comment above about test scripts setting `$(IOCSH_PS1)` (yes, they should always do that if they match the IOCSH prompt string to find the end of a command) I realize that trying to remove the color macros such as `ANSI_GREEN(…)` from the `configure/CONFIG_SITE_ENV` default value could break those scripts.

I have tightened up the bldEnvData.pl script to syntax-check the default values while still allowing `ANSI_ENV_<color>` and `ANSI_<color>("<string>")` macros in those default values; syntax errors in the parameter values should now halt the build at that script, not at the subsequent compilation of the envData.c file.

mdavidsaver commented:
With this PR, there are a lot more ANSI escapes going to stdout/err.  So much so that a capture via `tee` is hard to parse by eye.  However, I find that `less -R` will decode the color escapes nicely.

```sh
$ softIoc some.cmd 2>&1 | tee log
...

$ less -R log
```

anjohnson commented:
@mdavidsaver Typo, `less -R log`?

I have `$LESS` set to `-R -F -X`, although I rarely send IOC output to a log file. Is something like this worth adding to the Release Notes?

mdavidsaver commented:
> Typo, `less -R log`?

corrected.
